### ISSUE TITLE
Pruning commented-out code in library?

### ISSUE THREAD

**Issue by peterlefanulumsdaine** [UniMath/UniMath#1347]
Looking through the library during the refactorings I’ve been doing, I’ve ended up coming across quite a few sections of code that have been commented out and left to rot; for example, the 40 lines beginning here: https://github.com/UniMath/UniMath/blob/8065a80010b3637d5850c8ea2731110299ea1c6b/UniMath/CategoryTheory/limits/terminal.v#L123-L130 

Commenting out code can often be useful in the short-term (e.g. during WIP, when an upstream changes breaks compilation and it takes time to fix the downstream code), and can occasionally make sense in the long term (e.g. showing an alternative approach for a proof), but it’s generally bad to keep it around in the long term (mainly because it doesn’t get checked, so will rot over time).

**Is it OK to clean up such commented-out code when I come across it?**  My inclination is something like:

- when it seems like the commenting-out was because someone wanted to delete the code but keep it recoverable: then actually delete it (since it’s always recoverable from repo history anyway)
- when it’s because the code would be nice to have, but broke at some point and didn’t get fixed: then attempt to fix it; if that’s too hard/time-consuming, either delete it, or (if the material seems valuable and difficult to re-write) leave it but add comments documenting the situation
- when it’s there for another good reason (e.g. documenting an alternative approach), try to change to a form where it’s live instead of commented-out (e.g. with `Lemma mylemma_alternative […] Abort.`), or if that’s difficult/undesirable, leave it but make sure the intention is clearly documented in comments.

Does that sound a good approach? Or should we be more cautious or more aggressive about clearing out such code?

**Comment by benediktahrens** [comment]
The guidelines that you suggest sound very good to me; classifying a piece of commented-out code according to them might not be straightforward.

NB: Recovering deleted code from a repository seems to be to require a better memory than I have (not sure about you!), so this seems unrealistic to me in practice. I would only delete stuff if I were sure never to be interested in it again.

**Comment by peterlefanulumsdaine** [comment]
> classifying a piece of commented-out code according to them might not be straightforward.

Absolutely, agreed :-)  I wasn’t meaning to suggest them as instructions — more just to establish guidelines that it’s OK to use our own judgement within.

> Recovering deleted code from a repository seems to be to require a better memory than I have (not sure about you!)

Again, definitely agreed.  Discovering code that’s still live within the library is hard enough already!  Bringing the code up to date to compile is definitely a better option when possible, if the code is at all useful.

**Comment by DanGrayson** [comment]
I agree with the points that have been made.
